File: src/main/java/com/herbertgao/telegram/database/service/ExamDateService.java
package com.herbertgao.telegram.database.service;

import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper;
import com.herbertgao.telegram.database.entity.ExamDate;
import com.herbertgao.telegram.database.mapper.ExamDateMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;

/**
 * @program: gaokao_bot
 * @description: ExamDateService
 * @author: HerbertGao
 * @create: 2019-06-09 01:37
 **/
@Service
public class ExamDateService {

    @Autowired
    private ExamDateMapper mapper;

    /**
     * @param now
     * @return
     */
    public List<ExamDate> getExamList(LocalDateTime now, boolean desc) {
        QueryWrapper<ExamDate> wrapper = new QueryWrapper<>();
        wrapper.lambda()
                .le(ExamDate::getExamYearBeginDate, now)
                .ge(ExamDate::getExamYearEndDate, now)
                .eq(ExamDate::getIsDelete, false);
        if (desc) {
            wrapper.lambda().orderByDesc(ExamDate::getExamYearBeginDate);
        }
        return mapper.selectList(wrapper);
    }

    /**
     * @param year
     * @return
     */
    public List<ExamDate> getExamByYear(Integer year) {
        QueryWrapper<ExamDate> wrapper = new QueryWrapper<>();
        wrapper.lambda()
                .eq(ExamDate::getExamYear, year)
                .eq(ExamDate::getIsDelete, false);
        return mapper.selectList(wrapper);
    }
}

File: internal/service/exam_date_service.go
package service

import (
	"time"

	"github.com/herbertgao/gaokao_bot/internal/model"
	"github.com/herbertgao/gaokao_bot/internal/repository"
)

// ExamDateService 考试日期服务
type ExamDateService struct {
	repo *repository.ExamDateRepository
}

// NewExamDateService 创建考试日期服务
func NewExamDateService(repo *repository.ExamDateRepository) *ExamDateService {
	return &ExamDateService{repo: repo}
}

// GetExamsInRange 获取时间范围内的考试
func (s *ExamDateService) GetExamsInRange(now time.Time) ([]model.ExamDate, error) {
	return s.repo.GetExamsInRange(now)
}

// GetExamByYear 按年份获取考试
func (s *ExamDateService) GetExamByYear(year int) ([]model.ExamDate, error) {
	return s.repo.GetExamByYear(year)
}

// GetNextExamDate 获取下一个高考日期
func (s *ExamDateService) GetNextExamDate() (*model.ExamDate, error) {
	now := time.Now()
	exams, err := s.repo.GetExamsInRange(now)
	if err != nil {
		return nil, err
	}

	if len(exams) == 0 {
		return nil, nil
	}

	// 返回第一个（最近的）考试
	return &exams[0], nil
}
